### DISCUSSION TITLE
Command line based debugger for C# running on .NET 6

### DISCUSSION THREAD

**Discussion by awa5114** [dotnet/runtime#70440] [Q&A]
[Crossposting from SE](https://stackoverflow.com/questions/72548721/debug-c-sharp-from-the-command-line)
[Crossposting from Roslyn discussions](https://github.com/dotnet/csharplang/discussions/6203#discussion-4128667)

I'm on a (possibly futile) quest to find a command line based debugger for C# applications running on the .NET 6 framework. Something similar to `jdb` for java or `pdb` for python. I'm on a windows machine. Here are the meagre resources I have found:

- [debugging c# in the command line](https://stackoverflow.com/questions/1656864/debugging-c-sharp-in-the-command-line)
- [MDbg](https://stackoverflow.com/questions/2791459/where-can-i-get-mdbg)
- [DebugExe flag for `devenv`](https://docs.microsoft.com/en-us/visualstudio/ide/reference/debugexe-devenv-exe?view=vs-2022)

The most promising by far would have been `MDbg`, but this looks very much outdated. [The page on the NuGet repository][1] shows no version compatible with .NET 6.

I also made a [post][2] on reddit  but this unfortunately did not generate much discussion or answers.

Being able to hop into the debugger from the command line is a common feature among several languages, so I feel like it should exist for C#, however convoluted the process may be.

Is there any way to do this in C#?

  [1]: https://www.nuget.org/packages/MDbg/
  [2]: https://www.reddit.com/r/csharp/comments/uny9jn/command_line_based_debugger_for_csharp/

**Comment by alexrp** [comment]
I believe Samsung's NetCoreDbg has a CLI and runs on Windows: https://github.com/Samsung/netcoredbg

I haven't used it so I can't say much beyond that.

**Reply by awa5114** [reply]
It looks like a promising project but lacks a bit of maturity as far as I can tell. It doesn't look like an easy install, I have to build the code from source, etc which is beyond my skill level and time available at this point.

Hard to believe this is not a readily available feature for the .NET runtime. What happened to MSdebug? Why was it discontinued?

**Reply by alexrp** [reply]
By far the biggest impediment is the fact that the managed library that makes it easy to talk to the CoreCLR debugger APIs is not open source: https://github.com/dotnet/core/issues/505

The good news is that the low-level APIs are just a small native shim library + plain old (nano-)COM interfaces. Talking to them is actually a lot easier than people seem to think; it just involves writing a lot of boring COM interop glue. I was in the process of doing this, but decided to put the project on hold until this is implemented: https://github.com/dotnet/runtime/issues/66674

**Reply by PathogenDavid** [reply]
> I have to build the code from source

@awa5114 Maybe I'm missing something, but it appears that they have pre-built binaries attached to each release. ([For example: 2.0.0-915](https://github.com/Samsung/netcoredbg/releases/tag/2.0.0-915))

**Reply by awa5114** [reply]
@alexrp Thanks for linking to those threads. I finally see I'm not the only one struggling with this.

It looks like my only option for the time being is perhaps to give the Samsung CLI a try.

I'm not a .NET expert, but I would be open to lending a hand with the grunt work required to build a proper command line based debugger. Let me know if this might come in handy.

**Reply by awa5114** [reply]
@PathogenDavid strange that they don't include this in their installation instructions...

**Reply by PathogenDavid** [reply]
@awa5114 It looks like this is actually the first release where they've published binaries, but either way attaching binaries to releases on GitHub is fairly standard practice.

**Reply by awa5114** [reply]
@PathogenDavid I got it to work. Thanks.

**Reply by awa5114** [reply]
It looks like the CLI for `netcoredbg` is broken. I tried it for a simple console app and the developers confirmed that it didn't work. https://github.com/Samsung/netcoredbg/issues/92

**Reply by danmoseley** [reply]
Cc @tommcdon

**Reply by tommcdon** [reply]
One quick note - there are no known API changes that would break backwards compatibility for debuggers; however, `dbgshim.dll` has been moved from the dotnet/runtime repo in .net 7 to dotnet/diagnostics.  Please see https://github.com/dotnet/runtime/pull/67725, https://github.com/dotnet/diagnostics/pull/2842, and https://github.com/dotnet/diagnostics/pull/3088.  Dbgshim packages can be downloaded from https://github.com/dotnet/diagnostics/releases/tag/v6.0.328102 or from nuget.org.  The package names are `https://www.nuget.org/packages/Microsoft.Diagnostics.DbgShim.linux-<arch>`.  It's possible `netcoredbg` has a dependency on `dbgshim` and so copying the dbgshim module from the appropriate nuget package to the runtime install directory might resolve the issue.

**Reply by awa5114** [reply]
CC @viewizard  please see comment above by @tommcdon regarding windows compatibility for `netcoredbg`
